pkg/ignore/tokenizer.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/scanner"
+)
+
+// toFileIgnorePattern is used to perform normalization on the pattern like
+// - clean up the path to be good file path
+// - make sure paths start with /
+// - check if the patterns has inversions i.e !foo kind of things
+// - compile the pattern to regular expression
+func toFileIgnorePattern(pattern string) (*fileIgnorePattern, error) {
+
+	// clean the pattern to be well formed Go path
+	pattern = filepath.Clean(pattern)
+
+	// make sure the path starts with /
+	pattern = filepath.FromSlash(pattern)
+
+	ignorePattern := &fileIgnorePattern{}
+	ignorePattern.Pattern = pattern
+
+	// check if it has inverts and remove them before creating paths
+	if strings.HasPrefix(pattern, "!") {
+		pattern = strings.TrimPrefix(pattern, "!")
+		ignorePattern.Invert = true
+	}
+
+	// add the parent directories if and only if the pattern has a /
+	// otherwise the path is deemed to be under root
+	if strings.Contains(pattern, string(os.PathSeparator)) {
+		ptokens := strings.Split(pattern, string(os.PathSeparator))
+		ignorePattern.Paths = ptokens[:len(ptokens)-1]
+	}
+
+	re, err := asRegExp(pattern)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ignorePattern.RegexPattern = re
+
+	return ignorePattern, nil
+}
+
+// make each pattern as valid regular expression that can be compared
+// with file path
+func asRegExp(pattern string) (*regexp.Regexp, error) {
```

Comment:
I think that if this stopped just short of compiling the regular expression you'd have a simple `string -> string` function that would be really great for table testing.

Reply:
@mattmoor  - not sure I got you right, you mean to say to have the RegExp captured as string and do the compilation when it is used at https://github.com/kameshsampath/mink/blob/handle-ignore/pkg/ignore/types.go#L108 ? And make this function just take and return `string`   -- `func asRegExp(pattern string) string` -- ???